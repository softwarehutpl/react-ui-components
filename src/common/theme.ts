const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
