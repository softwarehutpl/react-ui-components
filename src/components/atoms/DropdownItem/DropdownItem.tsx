import React, { useContext } from 'react';
import styled from 'styled-components';
import { mapTextPositionToFlexJustify } from '../../../helpers/styleHelperMethods';
import DropdownContext from '../../../context/dropdownContext';
import { COLOR_ALTO } from '../../../common/constants/colors';

interface IDropdownItemProps {
  children?: React.ReactNode;
  className?: string;
  divider?: boolean;
  onClick?: () => void;
}

interface IDropdownItemStyleProps {
  fontColor?: string;
  backgroundColor?: string;
  height?: number;
  padding?: number;
  textAlignment?: 'left' | 'center' | 'right';
  heading?: boolean;
  dividerColor?: string;
  color?: string;
  disabled?: boolean;
}

const DropdownItem = ({ children, className, divider, onClick }: IDropdownItemProps) => {
  if (divider) {
    return <div className={className} />;
  }
  return (
    <div className={className} onClick={onClick}>
      {children}
    </div>
  );
};

const defaultProps = {
  children: null,
  className: '',
  onClick: () => {},
  height: 35,
  padding: 15,
  textAlignment: 'left',
  heading: false,
  divider: false,
  dividerColor: COLOR_ALTO,
  color: 'error',
  disabled: false,
};

DropdownItem.defaultProps = defaultProps;

const StyledDropdownItem = styled(DropdownItem)<IDropdownItemStyleProps>`
  ${({
    theme,
    color = defaultProps.color,
    fontColor,
    backgroundColor,
    height = defaultProps.height,
    padding = defaultProps.padding,
    textAlignment = defaultProps.textAlignment,
    heading = defaultProps.heading,
    divider = defaultProps.divider,
    dividerColor = defaultProps.dividerColor,
    disabled = defaultProps.disabled,
  }) => {
    if (divider) {
      return {
        height: '1px',
        'background-color': dividerColor,
      };
    }
    return {
      color: fontColor || theme.colors[color].base,
      'background-color': backgroundColor || theme.colors[color].light,
      width: '100%',
      height: `${height}px`,
      padding: `0 ${padding}px`,
      'box-sizing': 'border-box',
      display: 'flex',
      'align-items': 'center',
      'justify-content': mapTextPositionToFlexJustify(textAlignment),
      'font-size': heading ? '0.75em' : '1em',
      'font-weight': heading ? 'bold' : 'bolder',
      'text-transform': heading ? 'uppercase' : 'capitalize',
      '&:hover': {
        cursor: 'pointer',
        'background-color': theme.colors[color].dark,
      },
      'pointer-events': disabled ? 'none' : 'auto',
    };
  }}
`;

type DropdownItemProps = IDropdownItemStyleProps & IDropdownItemProps;

const withDropdownContext = (WrappedComponent: React.ComponentType<DropdownItemProps>) => (
  props: DropdownItemProps
) => {
  const { dropdownColor } = useContext(DropdownContext);
  const { color } = props;
  // if color is passed to item, render item with given color, if not, take color from dropdown
  return <WrappedComponent {...props} color={color || dropdownColor} />;
};

export default withDropdownContext(StyledDropdownItem);
