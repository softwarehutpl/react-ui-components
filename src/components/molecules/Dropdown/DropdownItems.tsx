import styled from 'styled-components';
import { defaultProps } from './Dropdown';

export interface IDropdownOptions {
  openingDirection?: 'down' | 'up';
  toggleHeight: number;
  itemsBackgroundColor?: string;
  color?: string;
}

export const DropdownOptions = styled.div<IDropdownOptions>`
  ${({
    openingDirection = defaultProps.openingDirection,
    toggleHeight,
    itemsBackgroundColor,
    theme,
    color = defaultProps.color,
  }) => ({
    position: `absolute`,
    top: openingDirection === 'down' ? `${toggleHeight + 5}px` : 'auto',
    bottom: openingDirection === 'up' ? `${toggleHeight + 5}px` : 'auto',
    left: 0,
    'border-radius': '5px',
    padding: '5px 0',
    'box-shadow': '0 4px 10px 4px rgba(217,217,217,0.6)',
    'background-color': itemsBackgroundColor || theme.colors[color].light,
    width: '100%',
  })}
`;
