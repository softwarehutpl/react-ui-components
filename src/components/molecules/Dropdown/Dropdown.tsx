import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import { DownArrow } from '@styled-icons/boxicons-solid/DownArrow';
import { UpArrow } from '@styled-icons/boxicons-solid/UpArrow';
import DropdownContext from '../../../context/dropdownContext';
import { DropdownToggle, IDropdownToggle } from './DropdownToggle';
import { DropdownOptions } from './DropdownItems';
import useOutsideClick from '../../../common/hooks/clickOutside';

interface IDropdown extends IDropdownToggle {
  children?: React.ReactNode;
  className?: string;
  title: string;
  itemsBackgroundColor?: string;
  itemsFontColor?: string;
  caretSize?: number;
  openingDirection?: 'down' | 'up';
}

interface IStyledDropdown {
  margin?: number;
}

const arrowStyle = (size?: number) => ({
  width: `${size}px`,
  marginLeft: '10px',
});

const Dropdown = ({
  children,
  color,
  className,
  title,
  width,
  itemsBackgroundColor,
  itemsFontColor,
  openingDirection,
  caretSize,
  disabled,
  backgroundColor,
  padding,
  fontColor,
}: IDropdown) => {
  const [isOpen, setIsOpen] = useState(false);
  const [toggleHeight, setToggleHeight] = useState(0);
  const toggleRef = React.useRef<HTMLDivElement>(null);
  const dropdownRef = React.useRef<HTMLDivElement>(null);

  const toggleDropdown = () => {
    if (!disabled) {
      setIsOpen(!isOpen);
    }
  };

  const closeDropdown = () => {
    setIsOpen(false);
  }

  useOutsideClick(dropdownRef, closeDropdown);

  useEffect(() => {
    if (toggleRef && toggleRef.current) {
      const { height } = toggleRef.current.getBoundingClientRect();
      setToggleHeight(height);
    }
  }, []);

  return (
    <DropdownContext.Provider
      value={{
        dropdownColor: color,
        dropdownWidth: width,
        itemsBackgroundColor,
        itemsFontColor,
        onDropdownItemClick: closeDropdown,
      }}
    >
      <div className={className} ref={dropdownRef}>
        <DropdownToggle
          ref={toggleRef}
          onClick={toggleDropdown}
          color={color}
          width={width}
          backgroundColor={backgroundColor}
          padding={padding}
          fontColor={fontColor}
          disabled={disabled}
        >
          {title}
          {openingDirection === 'up' ? (
            <UpArrow style={arrowStyle(caretSize)} />
          ) : (
            <DownArrow style={arrowStyle(caretSize)} />
          )}
        </DropdownToggle>
        {isOpen && (
          <DropdownOptions
            openingDirection={openingDirection}
            toggleHeight={toggleHeight}
            itemsBackgroundColor={itemsBackgroundColor}
            color={color}
          >
            {children}
          </DropdownOptions>
        )}
      </div>
    </DropdownContext.Provider>
  );
};

export const defaultProps = {
  color: 'primary',
  width: 200,
  padding: 15,
  margin: 0,
  openingDirection: 'down',
  caretSize: 15,
  disabled: false,
};

Dropdown.defaultProps = defaultProps;

const StyledDropdown = styled(Dropdown)<IStyledDropdown>`
  ${({ margin = defaultProps.margin }) => ({
    margin: `${margin}px`,
    position: 'relative',
  })}
`;

export default StyledDropdown;
