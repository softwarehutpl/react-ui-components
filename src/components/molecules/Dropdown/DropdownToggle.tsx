import styled from 'styled-components';
import { defaultProps } from './Dropdown';

export interface IDropdownToggle {
  color?: string;
  width?: number;
  disabled?: boolean;
  backgroundColor?: string;
  padding?: number;
  fontColor?: string;
}

export const DropdownToggle = styled.div<IDropdownToggle>`
  ${({
    theme,
    color = defaultProps.color,
    width = defaultProps.width,
    backgroundColor,
    padding = defaultProps.padding,
    fontColor,
    disabled = defaultProps.disabled,
  }) => {
    let dropdownColor = color;
    if (disabled) {
      dropdownColor = 'disabled';
    }
    return {
      width: `${width}px`,
      'background-color': backgroundColor || theme.colors[dropdownColor].base,
      padding: `${padding}px`,
      color: fontColor || theme.colors[dropdownColor].light,
      'font-weight': 'bold',
      'box-sizing': 'border-box',
      '&:hover': {
        cursor: disabled ? 'default' : 'pointer',
      },
      'text-align': 'left',
    };
  }}
`;
