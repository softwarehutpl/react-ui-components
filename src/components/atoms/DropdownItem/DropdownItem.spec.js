import React from 'react';
import { create, act } from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import DropdownItem from './DropdownItem';
import theme from '../../../common/theme';
import Dropdown from '../../molecules/Dropdown/Dropdown';
import {DropdownToggle} from '../../molecules/Dropdown/DropdownToggle';

describe('DropdownItem component', () => {
  it('shoud render correctly', () => {
    const tree = create(<DropdownItem theme={theme}>item</DropdownItem>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render dropdown item with its child', () => {
    const component = create(<DropdownItem theme={theme}>item</DropdownItem>);
    const instance = component.root;
    const dropdownItem = instance.findByType('div');
    expect(dropdownItem.props.children).toBe('item');
  });

  it('should render dropdown item with given classes', () => {
    const component = create(
      <DropdownItem className="test-class" theme={theme}>
        item
      </DropdownItem>
    );
    const instance = component.root;
    const dropdownItem = instance.findByType('div');
    expect(dropdownItem.props.className).toEqual(expect.stringContaining('test-class'));
  });

  it('should render dropdown item with color context from dropdown after dropdown toggle click', () => {
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <Dropdown title="title" color="secondary">
          <DropdownItem>item</DropdownItem>
        </Dropdown>
      </ThemeProvider>
    );
    const dropdownToggle = getByText('title');
    fireEvent.click(dropdownToggle);
    const dropdownItem = getByText('item');
    expect(dropdownItem).toHaveStyle('background-color: #bbdefb')
    expect(dropdownItem).toHaveStyle('color: #2196f3');
  });

  it('should render dropdown item with custom colors after dropdown toggle click', () => {
    const { getByText } = render(
      <ThemeProvider theme={theme}>
        <Dropdown title="title" itemsBackgroundColor="yellow" itemsFontColor="red">
          <DropdownItem>item</DropdownItem>
        </Dropdown>
      </ThemeProvider>
    );
    const dropdownToggle = getByText('title');
    fireEvent.click(dropdownToggle);
    const dropdownItem = getByText('item');
    expect(dropdownItem).toHaveStyle('background-color: yellow');
    expect(dropdownItem).toHaveStyle('color: red');
  });

  it('should render dropdown item with custom height', () => {
    const component = create(
      <DropdownItem theme={theme} height={40}>
        item
      </DropdownItem>
    );
    const dropdownItem = component.toJSON();
    expect(dropdownItem).toHaveStyleRule('height', '40px');
  });

  it('should render dropdown item with custom padding', () => {
    const component = create(
      <DropdownItem theme={theme} padding={30}>
        item
      </DropdownItem>
    );
    const dropdownItem = component.toJSON();
    expect(dropdownItem).toHaveStyleRule('padding', '0 30px');
  });

  it('should render dropdown item with given text placement', () => {
    const component = create(
      <DropdownItem theme={theme} textAlignment="right">
        item
      </DropdownItem>
    );
    const dropdownItem = component.toJSON();
    expect(dropdownItem).toHaveStyleRule('justify-content', 'flex-end');
  });

  it('should render dropdown item with style proper for heading', () => {
    const component = create(
      <DropdownItem theme={theme} heading>
        item
      </DropdownItem>
    );
    const dropdownItem = component.toJSON();
    expect(dropdownItem).toHaveStyleRule('text-transform', 'uppercase');
  });

  it('should render dropdown item with style proper for divider', () => {
    const component = create(<DropdownItem theme={theme} divider />);
    const dropdownItem = component.toJSON();
    expect(dropdownItem).toHaveStyleRule('height', '1px');
    expect(dropdownItem).toHaveStyleRule('background-color', '#d9d9d9');
  });

  it('should call dropdown item onClick', () => {
    const mockOnClick = jest.fn();
    const component = create(<DropdownItem theme={theme} onClick={mockOnClick} />);
    const instance = component.root;
    const dropdownItem = instance.findByType(DropdownItem);
    act(() => {
      dropdownItem.props.onClick();
    });
    expect(mockOnClick.mock.calls.length).toEqual(1);
  });

  it('if disabled should be not clickable', () => {
    const component = create(<DropdownItem theme={theme} disabled />);
    const dropdownItem = component.toJSON();
    expect(dropdownItem).toHaveStyleRule('pointer-events', 'none');
  });
});
