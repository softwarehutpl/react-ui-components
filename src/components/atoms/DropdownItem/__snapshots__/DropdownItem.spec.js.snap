// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownItem component shoud render correctly 1`] = `
.c0 {
  color: #000;
  background-color: #fff;
  width: 200px;
  height: 35px;
  padding: 0 15px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  font-size: 1em;
  font-weight: 600;
  text-transform: capitalize;
  pointer-events: auto;
}

.c0:hover {
  cursor: pointer;
  background-color: #f6f6f6;
}

<div
  className="c0"
  onClick={[Function]}
>
  item
</div>
`;
